src/read/pipelining.rs

```diff
@@ -0,0 +1,1209 @@
+            #[cfg_attr(not(unix), allow(unused_variables))]
+            let AllocatedHandles {
+                file_handle_mapping,
+                perms_todo,
+            } = transform_entries_to_allocated_handles(td.path(), trie).unwrap();
+
+            let mut files: Vec<_> = file_handle_mapping.into_iter().collect();
+            assert_eq!(1, files.len());
+            let (file_data, mut file) = files.pop().unwrap();
+            assert_eq!(
+                file_data,
+                ZipDataHandle::wrap(zip.shared.files.get_index(0).unwrap().1)
+            );
+            /* We didn't write anything to the file, so it's still empty at this point. */
+            assert_eq!(b"", &fs::read(td.path().join("a/b/c")).unwrap()[..]);
+            file.write_all(b"asdf").unwrap();
+            file.sync_data().unwrap();
+            /* Now the data is synced! */
+            assert_eq!(b"asdf", &fs::read(td.path().join("a/b/c")).unwrap()[..]);
+
+            #[cfg(unix)]
+            assert_eq!(
+                perms_todo
+                    .into_iter()
+                    .map(|(path, perms)| (path, perms.mode() & 0o777))
+                    .collect::<Vec<_>>(),
+                vec![
+                    (td.path().join("a/b/c"), 0o765),
+                    (td.path().join("a/b"), 0o500),
+                ]
+            );
+        }
+    }
+}
+
+#[cfg(unix)]
+pub mod split_extraction {
+    use displaydoc::Display;
+    use thiserror::Error;
+
+    use std::fs;
+    use std::io;
+    use std::mem::{self, MaybeUninit};
+    use std::path::Path;
+    use std::sync::mpsc;
+    use std::thread;
+
+    use crate::compression::CompressionMethod;
+    use crate::read::ZipArchive;
+    use crate::read::{make_crypto_reader, make_reader};
+    use crate::result::ZipError;
+    use crate::spec::FixedSizeBlock;
+    use crate::types::{ZipFileData, ZipLocalEntryBlock};
+
+    #[cfg(not(target_os = "linux"))]
+    use crate::read::split::pipe::unix::{PipeReadBufferSplicer, PipeWriteBufferSplicer};
+    #[cfg(target_os = "linux")]
+    use crate::read::split::{
+        file::linux::FileCopy,
+        pipe::linux::{PipeReadSplicer, PipeWriteSplicer},
+    };
+    use crate::read::split::{
+        file::{
+            unix::{FileBufferCopy, FileInput, FileOutput},
+            CopyRange, InputFile,
+        },
+        pipe::{unix::create_pipe, ReadSplicer, WriteSplicer},
+        util::{copy_via_buf, TakeWrite},
+    };
+
+    use super::{
+        handle_creation::{
+            transform_entries_to_allocated_handles, AllocatedHandles, HandleCreationError,
+            ZipDataHandle,
+        },
+        path_splitting::{lexicographic_entry_trie, PathSplitError},
+    };
+
+    /// Errors encountered during the split pipelined extraction process.
+    #[derive(Debug, Display, Error)]
+    pub enum SplitExtractionError {
+        /// i/o error: {0}
+        Io(#[from] io::Error),
+        /// zip error: {0}
+        Zip(#[from] ZipError),
```

Comment:
Combine `Io` and `Zip` since `ZipError` already has a variant for `io::Error`. I recommend this even if you end up having to implement `From<io::Error>` by hand (which will happen if `#[from]` isn't transitive).

Reply:
I don't see a way to make `#[from]` transitive in the `thiserror` docs, so I simply merged the two cases and added another `From<io::Error>` impl as you proposed!